enclave/server.ts

```diff
@@ -123,23 +144,33 @@ async function login(
  * must respond to a leaf event.
  */
 async function getSignature(socket: Socket, uFrom: any, uTo: any) {
-    const uFromAsTile = Tile.fromJSON(uFrom);
-    const hUFrom = uFromAsTile.hash();
-    const uToAsTile = Tile.fromJSON(uTo);
-    const hUTo = uToAsTile.hash();
-
-    claimedMoves.set(hUFrom.concat(hUTo), {
-        uFrom: uFromAsTile,
-        uTo: uToAsTile,
-    });
+    const pubkey = idToPubKey.get(socket.id);
+    if (pubkey) {
+        // Players cannot make more than one move per block
+        const latestBlock = playerLatestBlock.get(pubkey);
+        if (latestBlock && latestBlock < currentBlockHeight) {
+            const uFromAsTile = Tile.fromJSON(uFrom);
+            const hUFrom = uFromAsTile.hash();
+            const uToAsTile = Tile.fromJSON(uTo);
+            const hUTo = uToAsTile.hash();
+
+            claimedMoves.set(hUFrom.concat(hUTo), {
+                uFrom: uFromAsTile,
+                uTo: uToAsTile,
+                blockSubmitted: currentBlockHeight,
+            });
 
-    const digest = utils.solidityKeccak256(
-        ["uint256", "uint256"],
-        [hUFrom, hUTo]
-    );
-    const sig = await signer.signMessage(utils.arrayify(digest));
+            const digest = utils.solidityKeccak256(
+                ["uint256", "uint256", "uint256"],
+                [currentBlockHeight, hUFrom, hUTo]
+            );
+            const sig = await signer.signMessage(utils.arrayify(digest));
 
-    socket.emit("getSignatureResponse", sig, uFrom, uTo);
+            socket.emit("signatureResponse", sig, currentBlockHeight);
+        } else {
+            socket.emit("errorResponse", "Cannot move more than once per tick");
```

Comment:
should cut the connection- this is a malicious client 

there should also be client-side code for limiting to one move per block

Reply:
I think if a mod accidentally submits +1 move in a block, they shouldn't be booted immediately, no?

Reply:
or do we be very strict with client-side code for limiting moves / block

Reply:
yes strict & have client side code 

this pattern's common- having an exact definition of honest client behavior and terminating at any deviation